File: benchmark_filters.py
#!/usr/bin/env python3
"""
Benchmark script for comparing point cloud filtering methods in Python.

This script benchmarks Statistical Outlier Removal and DBSCAN filtering
on point clouds using Open3D in Python. It measures and reports the
processing time for each method.
"""

import numpy as np
import open3d as o3d
import time
import os
import sys

def create_test_point_cloud(num_normal=1000, num_outliers=50):
    """
    Creates a test point cloud with normal points and outliers.
    
    Args:
        num_normal: Number of normal points to generate
        num_outliers: Number of outlier points to generate
        
    Returns:
        o3d.geometry.PointCloud: The generated point cloud with outliers
    """
    # Create normal points in 3D space, centered around origin
    points = np.random.rand(num_normal, 3) - 0.5  # Range: [-0.5, 0.5]
    
    # Create outlier points that are farther from the center
    outliers = (np.random.rand(num_outliers, 3) - 0.5) * 3  # Range: [-1.5, 1.5]
    
    # Combine normal points and outliers into a single array
    all_points = np.vstack([points, outliers])
    
    # Create Open3D point cloud object
    pcd = o3d.geometry.PointCloud()
    pcd.points = o3d.utility.Vector3dVector(all_points)
    
    return pcd

def load_point_cloud(filename):
    """
    Load a point cloud from a PCD file.
    
    Args:
        filename: Path to the PCD file
        
    Returns:
        o3d.geometry.PointCloud: The loaded point cloud
    """
    if not os.path.exists(filename):
        print(f"Error: File {filename} not found.")
        return None
        
    return o3d.io.read_point_cloud(filename)

def statistical_outlier_removal(pcd, nb_neighbors=50, std_ratio=1.0):
    """
    Apply Statistical Outlier Removal filter.
    
    Args:
        pcd: Input point cloud
        nb_neighbors: Number of neighbors to analyze (equivalent to meanK in PCL)
        std_ratio: Standard deviation ratio (equivalent to stddevMulThresh in PCL)
        
    Returns:
        tuple: (filtered point cloud, processing time in seconds)
    """
    start_time = time.time()
    
    # Apply Statistical Outlier Removal
    cl, ind = pcd.remove_statistical_outlier(nb_neighbors=nb_neighbors, 
                                            std_ratio=std_ratio)
    
    end_time = time.time()
    processing_time = end_time - start_time
    
    return cl, processing_time

def dbscan_clustering(pcd, eps=0.15, min_points=10):
    """
    Apply DBSCAN clustering for outlier removal.
    
    Args:
        pcd: Input point cloud
        eps: Maximum distance between points to be considered in the same cluster
        min_points: Minimum number of points required to form a cluster
        
    Returns:
        tuple: (filtered point cloud, processing time in seconds)
    """
    start_time = time.time()
    
    # Compute DBSCAN clustering
    labels = np.array(pcd.cluster_dbscan(eps=eps, min_points=min_points))
    
    # Filter out noise points (label -1)
    filtered_pcd = o3d.geometry.PointCloud()
    inlier_indices = np.where(labels >= 0)[0]
    filtered_pcd.points = o3d.utility.Vector3dVector(
        np.asarray(pcd.points)[inlier_indices])
    
    end_time = time.time()
    processing_time = end_time - start_time
    
    return filtered_pcd, processing_time

def benchmark(input_file=None, num_runs=5):
    """
    Benchmark both filtering methods.
    
    Args:
        input_file: Path to input PCD file (if None, generates a test cloud)
        num_runs: Number of benchmark runs to average
    """
    # Load or create point cloud
    if input_file and os.path.exists(input_file):
        pcd = load_point_cloud(input_file)
        print(f"Loaded point cloud from {input_file} with {len(pcd.points)} points")
    else:
        pcd = create_test_point_cloud()
        print(f"Generated test point cloud with {len(pcd.points)} points")
        
    # Save for C++ comparison if needed
    if not os.path.exists("input_cloud.pcd"):
        o3d.io.write_point_cloud("input_cloud.pcd", pcd)
        print("Saved point cloud to 'input_cloud.pcd'")
    
    # Parameters (matching the C++ implementation)
    nb_neighbors = 50  # meanK in PCL
    std_ratio = 1.0    # stddevMulThresh in PCL
    eps = 0.15         # cluster_tolerance in PCL
    min_points = 10    # min_cluster_size in PCL
    
    # Benchmark Statistical Outlier Removal
    sor_times = []
    for i in range(num_runs):
        _, time_taken = statistical_outlier_removal(pcd, nb_neighbors, std_ratio)
        sor_times.append(time_taken)
    
    avg_sor_time = sum(sor_times) / len(sor_times)
    print(f"\nStatistical Outlier Removal (Python):")
    print(f"  Parameters: nb_neighbors={nb_neighbors}, std_ratio={std_ratio}")
    print(f"  Average processing time: {avg_sor_time:.6f} seconds")
    
    # Benchmark DBSCAN Clustering
    dbscan_times = []
    for i in range(num_runs):
        _, time_taken = dbscan_clustering(pcd, eps, min_points)
        dbscan_times.append(time_taken)
    
    avg_dbscan_time = sum(dbscan_times) / len(dbscan_times)
    print(f"\nDBSCAN Clustering (Python):")
    print(f"  Parameters: eps={eps}, min_points={min_points}")
    print(f"  Average processing time: {avg_dbscan_time:.6f} seconds")

if __name__ == "__main__":
    input_file = sys.argv[1] if len(sys.argv) > 1 else None
    benchmark(input_file)

File: src/benchmark_filters.cpp
/**
 * @file benchmark_filters.cpp
 * @brief Benchmark program for comparing point cloud filtering methods in C++
 * 
 * This program benchmarks Statistical Outlier Removal and DBSCAN filtering
 * on point clouds using PCL in C++. It measures and reports the
 * processing time for each method.
 */

#include <iostream>
#include <chrono>
#include <vector>
#include <numeric>
#include <iomanip>
#include <pcl/io/pcd_io.h>
#include <pcl/point_types.h>
#include <pcl/filters/statistical_outlier_removal.h>
#include <pcl/segmentation/extract_clusters.h>
#include <pcl/kdtree/kdtree.h>

/**
 * Apply Statistical Outlier Removal filter and measure processing time
 * 
 * @param cloud Input point cloud
 * @param meanK Number of neighbors to analyze
 * @param stddev_mult Standard deviation multiplier threshold
 * @return Pair of filtered cloud and processing time in seconds
 */
std::pair<pcl::PointCloud<pcl::PointXYZ>::Ptr, double> 
applyStatisticalOutlierRemoval(const pcl::PointCloud<pcl::PointXYZ>::Ptr& cloud, 
                              int meanK, double stddev_mult) {
    // Start timing
    auto start_time = std::chrono::high_resolution_clock::now();
    
    // Create the filtering object
    pcl::StatisticalOutlierRemoval<pcl::PointXYZ> sor;
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_filtered(new pcl::PointCloud<pcl::PointXYZ>);
    
    sor.setInputCloud(cloud);
    sor.setMeanK(meanK);
    sor.setStddevMulThresh(stddev_mult);
    sor.filter(*cloud_filtered);
    
    // End timing
    auto end_time = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed = end_time - start_time;
    
    return {cloud_filtered, elapsed.count()};
}

/**
 * Apply DBSCAN clustering for outlier removal and measure processing time
 * 
 * @param cloud Input point cloud
 * @param cluster_tolerance Maximum distance between points in a cluster
 * @param min_cluster_size Minimum number of points in a cluster
 * @param max_cluster_size Maximum number of points in a cluster
 * @return Pair of filtered cloud and processing time in seconds
 */
std::pair<pcl::PointCloud<pcl::PointXYZ>::Ptr, double> 
applyDBSCAN(const pcl::PointCloud<pcl::PointXYZ>::Ptr& cloud,
           double cluster_tolerance, int min_cluster_size, int max_cluster_size) {
    // Start timing
    auto start_time = std::chrono::high_resolution_clock::now();
    
    // Create result cloud
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_filtered(new pcl::PointCloud<pcl::PointXYZ>);
    
    // Create KdTree for searching
    pcl::search::KdTree<pcl::PointXYZ>::Ptr tree(new pcl::search::KdTree<pcl::PointXYZ>);
    tree->setInputCloud(cloud);
    
    // Perform DBSCAN clustering
    std::vector<pcl::PointIndices> cluster_indices;
    pcl::EuclideanClusterExtraction<pcl::PointXYZ> ec;
    ec.setClusterTolerance(cluster_tolerance);
    ec.setMinClusterSize(min_cluster_size);
    ec.setMaxClusterSize(max_cluster_size);
    ec.setSearchMethod(tree);
    ec.setInputCloud(cloud);
    ec.extract(cluster_indices);
    
    // Extract all points belonging to clusters (non-outliers)
    for (const auto& cluster : cluster_indices) {
        for (const auto& idx : cluster.indices) {
            cloud_filtered->push_back((*cloud)[idx]);
        }
    }
    
    // End timing
    auto end_time = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed = end_time - start_time;
    
    return {cloud_filtered, elapsed.count()};
}

/**
 * Main function for benchmarking point cloud filtering methods
 * 
 * @param argc Number of command line arguments
 * @param argv Command line arguments (optional: input PCD file path)
 * @return Exit code
 */
int main(int argc, char** argv) {
    // Load the input point cloud
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud(new pcl::PointCloud<pcl::PointXYZ>);
    
    // Check if a file path was provided
    std::string input_file = "input_cloud.pcd";
    if (argc > 1) {
        input_file = argv[1];
    }
    
    // Load the PCD file
    if (pcl::io::loadPCDFile<pcl::PointXYZ>(input_file, *cloud) == -1) {
        PCL_ERROR("Couldn't read file %s\n", input_file.c_str());
        return -1;
    }
    
    std::cout << "Loaded " << cloud->size() << " points from " << input_file << std::endl;
    
    // Parameters (matching the Python implementation)
    int meanK = 50;                // nb_neighbors in Python
    double stddev_mult = 1.0;      // std_ratio in Python
    double cluster_tolerance = 0.15; // eps in Python
    int min_cluster_size = 10;     // min_points in Python
    int max_cluster_size = 10000;  // not directly used in Python
    
    // Number of benchmark runs
    const int num_runs = 5;
    
    // Benchmark Statistical Outlier Removal
    std::vector<double> sor_times;
    for (int i = 0; i < num_runs; ++i) {
        auto result = applyStatisticalOutlierRemoval(cloud, meanK, stddev_mult);
        double time_taken = result.second;
        sor_times.push_back(time_taken);
    }
    
    double avg_sor_time = std::accumulate(sor_times.begin(), sor_times.end(), 0.0) / sor_times.size();
    std::cout << "\nStatistical Outlier Removal (C++):" << std::endl;
    std::cout << "  Parameters: meanK=" << meanK << ", stddev_mult=" << stddev_mult << std::endl;
    std::cout << "  Average processing time: " << std::fixed << std::setprecision(6) << avg_sor_time << " seconds" << std::endl;
    
    // Benchmark DBSCAN Clustering
    std::vector<double> dbscan_times;
    for (int i = 0; i < num_runs; ++i) {
        auto result = applyDBSCAN(cloud, cluster_tolerance, min_cluster_size, max_cluster_size);
        double time_taken = result.second;
        dbscan_times.push_back(time_taken);
    }
    
    double avg_dbscan_time = std::accumulate(dbscan_times.begin(), dbscan_times.end(), 0.0) / dbscan_times.size();
    std::cout << "\nDBSCAN Clustering (C++):" << std::endl;
    std::cout << "  Parameters: cluster_tolerance=" << cluster_tolerance 
              << ", min_cluster_size=" << min_cluster_size 
              << ", max_cluster_size=" << max_cluster_size << std::endl;
    std::cout << "  Average processing time: " << std::fixed << std::setprecision(6) << avg_dbscan_time << " seconds" << std::endl;
    
    return 0;
}
